Fix the keyboard overlapping buttons on the sign in form

tgolen opened the pull request:
### Details
The page was originally built with using a top margin set to a percentage of the page height. Because of that, when the keyboard opens, the page wasn't properly collapsing the top margin in order to reveal the entire form. I restructured the entire layout to properly use flex techniques to allow the section to collapse as intended.

### Fixed Issues
$ https://github.com/Expensify/App/issues/11701

### Tests
1. Be signed out
2. Have your email address entered so that you see the password form
3. For native platforms, verify the password form is not covered by the software keyboards
4. For all other platforms, verify that the form and page layout looks correct (best to compare it to production).

For web specifically:
1. Test the page at widths less than 800 (form should always be centered and have a max width of 400px. The form should NOT span the entire width of the page for widths between 400 and 800). 
2. Test the page at widths between 801 and 1024 (the form should no longer take up the top half of the page and it should look just like any width greater than 1024)
3. Test the page at widths greater than 1024 (form should be a fixed width of ~400px on the left side of the page)

Test keyboard dismissing on native (iOS and Android):
1. On both the enter email and enter password forms, tap into the text input
2. Tap outside of the input and verify the keyboard is dismissed and verify the input does not have focus (it does not have a blue border)
3. On the enter email form, enter the email of an existing account and either tap continue or press enter
4. Verify the keyboard remains shown on the password form and the password text input has focus
5. On the email form, enter an email address that doesn't have an account yet and press enter (or click continue)
6. Verify that you are shown the message that a link has been sent to you and verify the keyboard has been dismissed

- [ ] Verify that no errors appear in the JS console

### Offline tests
1. Go offline
2. Verify there is an offline indicator under the sign in button
3. Go online
4. Verify the offline indicator goes away

### QA Steps
Same as the tests above

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.
<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for expected offline behavior are in the `Offline steps` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots/Videos
<details>
<summary>Web</summary>

![2022-11-28_15-24-44](https://user-images.githubusercontent.com/1228807/204395183-a8aa5e9d-aa87-4f3e-b210-fdb4b5b3d931.png)
![image](https://user-images.githubusercontent.com/1228807/205118679-a6e23874-7b25-4d7f-97dd-739e854994d9.png)



</details>

<details>
<summary>Mobile Web - Chrome</summary>


![2022-11-28_15-25-57](https://user-images.githubusercontent.com/1228807/204395196-d87d9f86-39b3-4ffd-b5c8-9f756d50673f.png)
![image](https://user-images.githubusercontent.com/1228807/204867607-b46fde91-5d53-467b-96f8-6e81800c2a42.png)



</details>

<details>
<summary>Mobile Web - Safari</summary>

![2022-11-28_15-27-06](https://user-images.githubusercontent.com/1228807/204395213-970c2db9-ab98-460f-b22a-55664724c7be.png)

![image](https://user-images.githubusercontent.com/1228807/204867712-d1f623af-9993-4018-8696-f91b9285985c.png)


</details>

<details>
<summary>Desktop</summary>

![2022-11-28_15-27-50](https://user-images.githubusercontent.com/1228807/204395231-ac0dd8dc-4fa8-4a36-b77a-4923041bf476.png)


</details>

<details>
<summary>iOS</summary>


![2022-11-28_15-26-53](https://user-images.githubusercontent.com/1228807/204395243-75775e31-32fd-4103-af93-fc2518ff5138.png)
![image](https://user-images.githubusercontent.com/1228807/204867795-00dd3c43-73a3-4066-a642-90acb60345e0.png)



</details>

<details>
<summary>Android</summary>


![2022-11-28_15-25-03](https://user-images.githubusercontent.com/1228807/204395258-e7501ab6-e57b-4323-bc47-59d023a06001.png)
![image](https://user-images.githubusercontent.com/1228807/204867482-aae6dbeb-1fa5-448f-8c8a-e6175e522a66.png)



</details>

tgolen commented:
@rushatgabhane Yeah, feel free to review! I've gone and added you.

I have fixed the keyboard dismissing issue (had to add back in the scrollview).

tgolen commented:
Anyone can review it :D The more the merrier!

cristipaval reviewed (commented):
Hey @tgolen ! I tested your branch and looks like the submit button is still partially covered by the keyboard.


https://user-images.githubusercontent.com/18078393/203102735-4fa7224b-f58a-471b-adc6-8cf3774be389.mov


This is NexusOne with Android 10

cristipaval commented:
> @cristipaval could you please re-upload the recording, thanks!

should work now

tgolen commented:
Thanks @cristipaval! I think that will be OK. I don't think that's a supported device or android version? Is that what you normally test with?

cristipaval commented:
This is an emulator on which I usually test to see how the App behaves on smaller screens. The Android 10 it's not old I'd say, so we should support it.

I do agree that the behaviour it's still fine, even if the button is completely covered by the keyboard, as long as we have the submit action on the keyboard working as a physical tap on the submit button.

tgolen commented:
I checked with QA and they are only required to test on Android 11 and above, so I think what you're seeing there is OK.

cristipaval commented:
Yes it works better on newer Android versions.

Looks like `howsVerticalScrollIndicator={false}` does not work on my Pixel 5 emulator with Android 12.

https://user-images.githubusercontent.com/18078393/203124140-bf1ff078-34ca-4058-9955-d5c0f647f1e5.mov

tgolen commented:
OK, thanks for the testing! I'll work on those today and tomorrow and see what I come up with.

tgolen commented:
Oh, ha... typo in the prop for `howsVerticalScrollIndicator` 🤦

jasperhuangg commented on src/pages/signin/SignInPageLayout/SignInPageContent.js:
```diff
@@ -31,69 +29,60 @@ const propTypes = {
-                styles.flex1,
-                styles.signInPageNarrowContentContainer,
-            ]}
+        <ScrollView
+            style={[styles.h100]}
+            contentContainerStyle={[scrollViewContentContainerStyles]}
+            keyboardShouldPersistTaps="handled"
+            howsVerticalScrollIndicator={false}
+        >
+
+            {/* This empty view creates margin on the top of the sign in form which will shrink and grow depending on if the keyboard is open or not */}
```
Hi @tgolen, thought it was important to note while investigating https://github.com/Expensify/App/issues/11277 I found removing this `ScrollView` fixes any of its `KeyboardAvoidingView` children that aren't functioning as expected.

I am a bit confused at to why we used this `ScrollView` in the first place; based on the expected behavior from https://github.com/Expensify/App/issues/11277 we never want this page to become scrollable, so why use a `ScrollView`?

jasperhuangg replied:
Sorry, got ahead of myself, just saw this: https://github.com/Expensify/App/pull/12848#issuecomment-1320503333, which explains why the `ScrollView` is necessary.

tgolen replied:
Yeah, I definitely went back and forth about the scroll view. Maybe there is a better way to allow the keyboard to be dismissed than without a scroll view. Any ideas?

jasperhuangg replied:
We can use a `TouchableWithFeedback` and hook that up to `Keyboard.dismiss()`

tgolen replied:
Like, it's weird that the input wouldn't just blur when you tap outside of it, right?

jasperhuangg replied:
https://reactnative.dev/docs/keyboard#dismiss

tgolen replied:
I think the answer lies somewhere with this logic: I need to dig into it after lunch. https://github.com/Expensify/App/blob/0bbc47bbde1c2cda8ee4814ecd7f90c26d2f32f0/src/pages/signin/LoginForm.js#L91-L93

jasperhuangg replied:
```
function componentDidMount() {
    Keyboard.addListener('keyboardDidShow', () => this.setState({isKeyboardShowing: true}));
    Keyboard.addListener('keyboardDidHide', () => this.setState({isKeyboardShowing: false}));
}

function dismissKeyboardIfOpen() {
    if (!this.state.isKeyboardShowing) {
        return;
    }
    Keyboard.dismiss();
}
```

I'm going to try these out with what's still on `main` and but without the `ScrollView`.

If we decide to go this route it's probably cleanest to stick this in some HOC that has an index.ios.js file so we only use this logic for iOS.

tgolen replied:
Why would we only want it for iOS? If it works for other platforms, then that's fine, isn't it?

tgolen replied:
OK, Great! Then we just need to ensure this is the end result:

> Keyboard should stay active when transitioning to PasswordForm, but it should be dismissed when transitioning to ResendValidationForm

As well as allowing the inputs to be blurred when tapping outside of them (which should also hide the keyboard).

jasperhuangg replied:
https://github.com/Expensify/App/blob/0bbc47bbde1c2cda8ee4814ecd7f90c26d2f32f0/src/pages/signin/LoginForm.js#L91-L93

@tgolen I looked into the above logic wasn't able to see how we could use it to dismiss the keyboard. Removing it didn't allow tapping away from the keyboard, and I can't see how `componentDidUpdate` in that component would be invoked unless we somehow hooked up props/state somewhere to an `onPress` handler.

jasperhuangg replied:
Separately, I created a HOC `PressableDismissKeyboard` in this PR: https://github.com/Expensify/App/pull/12903:
- allows you to dismiss the keyboard on iOS. It basically does the hooking up of props/state to an `onPress` handler
- to use it you can change the `ScrollView` to a regular `View` and wrap the entire `SignInPageContent` component with it. Please let me know what you think about this

tgolen replied:
Thanks, Jasper! I think that component idea is neat. I just can't help thinking though that it's covering up the root of the problem. There must be _something_ that's causing those inputs to not be able to lose focus. Any other normal form with text inputs works (by "works" I mean that tapping outside a text input dismisses the keyboard).

Let's keep digging!

tgolen replied:
OK, I got this working, using your idea, Jasper. 

When I researched this, I found that I made a bad assumption. React-Native doesn't automatically dismiss the keyboard or blur an input when tapping outside of it. All the references I could find showed that it had to be manually done.

For now, I've only implemented this in the sign in form for now as I didn't want to cause a bunch of regressions in all the other forms.

jasperhuangg commented on src/pages/signin/SignInPageLayout/SignInPageContent.js:
```diff
@@ -28,47 +26,42 @@ const propTypes = {
-                <SignInPageForm style={[
+const SignInPageContent = (props) => {
+    const dismissKeyboardWhenTappedOutsideOfInput = () => {
+        // This prop comes from with
```
```suggestion
        // This prop comes from withKeyboardState
```

jasperhuangg commented on src/pages/signin/SignInPageLayout/SignInPageContent.js:
```diff
@@ -28,47 +26,42 @@ const propTypes = {
-                    props.isSmallScreenWidth ? styles.ph5 : styles.ph4,
+                    styles.signInPageLeftContainer,
+
+                    // Restrict the width if the left container only for large screens. For smaller screens, the width needs to be fluid to span the entire width of the page.
```
```suggestion
                    // Restrict the width of the left container only for large screens. For smaller screens, the width needs to be fluid to span the entire width of the page.
```

jasperhuangg commented on src/pages/signin/SignInPageLayout/SignInPageContent.js:
```diff
@@ -28,47 +26,42 @@ const propTypes = {
+                    styles.signInPageLeftContainer,
+
+                    // Restrict the width if the left container only for large screens. For smaller screens, the width needs to be fluid to span the entire width of the page.
+                    !props.isMediumScreenWidth && !props.isSmallScreenWidth && styles.signInPageWideLeftContainer,
```
NAB: Wondering why we don't have a `isLargeScreenWidth` prop for times like these

tgolen replied:
Actually, that's a great point. I want to refactor it a little to have this.

jasperhuangg reviewed (dismissed):
just noticed some typos, overall looks good

jasperhuangg commented:
@rushatgabhane @tgolen 

agree, let's leave some more margin at the top before the status bar, there's still some gap between the ToS and the keyboard that we can use

tgolen committed:
Refactor to include isLargeScreenWidth

tgolen committed:
Restrict the height of the spacer

tgolen commented:
OK, I've updated again to address PR comments and increase the margin at the top of the form to have a minHeight of 24px.

@rushatgabhane I _think_ it's fine the way it is showing the TOS. I looked at `main` and this is the behavior (for android):

1. tap into email field
2. TOS is shown under the keyboard
3. start typing
4. as soon as one key is pressed, then the TOS all of a sudden shift up to be above the keyboard

So, I actually think that this is kind of solving a bug (with the TOS jumping).

However, it's worth noting that on iOS, due to how the page adjusts to the keyboard being opened, there is different behavior than android and the TOS will always remain at the bottom of the page.

I still think that's OK and not worth chasing at this time (since you can still access the TOS) and some difference in platform behavior is expected and doesn't cause any UX issue.

rushatgabhane reviewed (changes requested):
> I think it's fine the way it is showing the TOS.

@tgolen quick thought: im not sure if i agree with showing the ToS above the keyboard.
Maybe it's just me 🤷‍♀️but I find it too distracting. 

The screen would look too cramped on a small device


<table>
<tr><td>Before vs After (**Ignore the offcenter in 2nd image**)</td></tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/29673073/203411893-0107f44e-43f8-4b5a-9c81-c6ece9c815a4.jpeg" height=500>
<img src="https://user-images.githubusercontent.com/29673073/203411901-eac5ea85-c741-4a92-bea1-5a661246b93a.jpeg" height=500>
</td>

  </tr>
</table>

aimane-chnaif reviewed (changes requested):
@tgolen I don't think screen looks good in middle screen width (=1024px) than before
Before:
<img width="1024" alt="before" src="https://user-images.githubusercontent.com/96077027/203413863-51e97052-68e4-4b69-a79d-dbc7f7e141f1.png">

After:
<img width="1024" alt="after" src="https://user-images.githubusercontent.com/96077027/203413925-cdf5b6a1-87e6-48b0-a5f0-ba803a12d58e.png">

width of input, button components became larger which look good but TOS part is too long

I am very familiar with it since I already worked on device responsive in signin page layout while fixing https://github.com/Expensify/App/issues/9817

tgolen commented on src/pages/signin/SignInPageLayout/SignInPageContent.js:
```diff
@@ -28,47 +26,42 @@ const propTypes = {
+                    styles.signInPageLeftContainer,
+
+                    // Restrict the width of the left container only for large screens. For smaller screens, the width needs to be fluid to span the entire width of the page.
+                    props.isLargeScreenWidth && styles.signInPageWideLeftContainer,
```
OK, Thanks. I can look into this. I am guessing that maybe the device width is greater than the breakpoint for isSmallScreenWidth? but I'll need to investigate. If it's past that width, we'll need to discuss how we want to solve that.

tgolen replied:
Ah, no. I see this on the pixel 4 too. I think it was the change I made to the padding. Let me think about this

tgolen commented:
You're right, that TOS stuff I didn't intend to span like that and I can make it the same width as the form.

shawnborton commented:
I like the idea, but it looks like the form area is too skinny now? Here is a comparison with how it currently looks:
<img width="374" alt="image" src="https://user-images.githubusercontent.com/2319350/203430838-762dc7b3-82eb-4111-9bc1-fbdb5555b991.png">

shawnborton commented:
It looks like we have too much padding on the left/right of the form now.

tgolen committed:
Make terms have a minimum width like the form

tgolen commented:
Thanks Shawn! I think I've mostly got it nailed down. Web looks fine, but iOS and Android has the form off center, but the width looks correct. I'll keep working on this, but it'll need to take a break while I am OOO for the rest of the week.

## Web
Padding is correct:

![image](https://user-images.githubusercontent.com/1228807/203439708-4d68f1c4-c26e-4705-800d-3c27c0821a5d.png)

## Android and iOS
Width of the form looks correct, but the horizontal spacing is off a little bit:

![image](https://user-images.githubusercontent.com/1228807/203440642-32620584-371d-4acf-87d2-a956ce274cf1.png)

shawnborton commented:
Sounds good! Keep me posted when this is ready for another review.

tgolen committed:
Merge branch 'main' into tgolen-fix-signin-keyboard-avoiding

tgolen committed:
Fix conflicts from theme work

tgolen commented:
The new theme changes that were merged caused a lot of my styles to be broken. I'll be spending some time fixing those and then we'll need to retest everything.

tgolen committed:
Get layout working on web and android

tgolen committed:
Remove medium screen layout

tgolen committed:
Remove medium screen styles

tgolen commented:
Alllrighty, I pushed a pretty big update to this PR. All the styles are looking very consistent. I'd appreciate some eagle-eyes of @shawnborton to take another look!

I updated the screenshots in the OG description and also removed the medium screen styles from web.

shawnborton commented:
Nice, this looks good to me! I guess the only thing we might want to see is a screenshot for mobile where you show that we've fixed the bug, basically showing where we fixed the keyboard overlapping the buttons?

tgolen commented:
OK, I've added screenshots for the 4 native platforms with keyboards to show the password form and that the sign in button is all visible.

@jasperhuangg @aimane-chnaif @rushatgabhane @cristipaval Please re-review this PR soon so it can get merged hopefully by the end of the week.

jasperhuangg reviewed (dismissed):
code LGTM and tests well 👍

shawnborton commented:
It looks like the horizontal padding increases when the keyboard is up on iOS?

No keyboard | keyboard
----- | -----
![image](https://user-images.githubusercontent.com/2319350/204620838-3f978b24-d7dc-42d0-a868-9438830b6b04.png) | ![image](https://user-images.githubusercontent.com/2319350/204620808-4f449496-7222-4a89-89cb-ba6119dd9ef5.png)

The padding is correct in the first screenshot.

tgolen committed:
Merge branch 'main' into tgolen-fix-signin-keyboard-avoiding

tgolen committed:
Removed unused import

tgolen committed:
Make sure the padding is the same on the password form

tgolen committed:
Add a little vertical mergin

tgolen commented:
Thanks again for all the testing. I know this one is a pain. It seems like such little changes can have very unexpected effects. I've gone through the reported issues:

- The horizontal padding on the password from is wrong

**FIXED** 

- Adding some padding above the "Not X? Go Back" text

**POSSIBLY FIXED** I wasn't able to reproduce this with my pixel 4a emulator, but I have changed it to have some margin `mv5` to hopefully help for whatever android client is in your screenshot.

- The keyboard "glitch" where the keyboard hides and then reshows quickly when going from the username to password form

**IGNORED** I was not able to solve this. I went so far as to remove every usage of `Keyboard.dismiss()` in our app, and I couldn't find anything that was causing the keyboard to dismiss. It's possible it's happening as part of our navigation library. It is also happening in `main` so it's not a regression in this PR and I think it's OK to merge.

- The extra padding in mWeb Safari at the bottom of the page

**IGNORED** I wasn't able to fix this one either. I see the padding is also there on `main`, so this is something else that isn't a regression in this PR. I think it has to do with the browsers address bar as seen in this video:


https://user-images.githubusercontent.com/1228807/204822382-59b02b20-8b23-4f27-9dfc-1b1b5c0bc177.mp4

tgolen commented:
Ah, yeah. At some point, there is only so much we can do with the height we are given. I agree it seems acceptable for now.

tgolen commented:
> Sounds good - just wanted to confirm if screenshots were updated or not?

@shawnborton They have now been updated, yes 😁

shawnborton commented:
For this size here: 
<img width="393" alt="image" src="https://user-images.githubusercontent.com/2319350/204885292-c19597cc-2cc3-4390-bda8-316aa0a4ff17.png">


I think it might look better to go back to what we originally had, where we give the form a max-width and just center it on the page:
<img width="731" alt="image" src="https://user-images.githubusercontent.com/2319350/204885403-fead0ec6-8e5b-47c1-ae22-b082ca00d81f.png">


Thoughts on getting that back in here?

tgolen commented:
OK, I updated that page size to have a fixed width form and updated the screenshots in the original description:

![image](https://user-images.githubusercontent.com/1228807/204902977-c0776629-dc19-4a14-9e55-bb260d8f144b.png)

shawnborton commented:
Awesome, thank you Tim!

tgolen committed:
Keep the form a fixed width on small to medium screens

tgolen commented:
Ack, I just realized I didn't push up the code for that last change. Anyway, the code is there now 😅

tgolen committed:
Center the form better and fix iphone SE

tgolen committed:
Always keep the form the same width

tgolen commented on src/styles/styles.js:
```diff
@@ -893,60 +893,27 @@ const styles = {
+    signInPageLeftContainerMediumScreen: {
+        maxWidth: 400,
+        marginLeft: 'auto',
+        marginRight: 'auto',
```
Ugh, this stuff is so finicky.

The reason I added the auto-margins is:

1. Shawn requested to have the form use a max-width
2. Setting a maxWidth works for the form, but for medium-sized screens, the form is aligned on the left side of the container and looks bad. The form needs to be horizontally centered in the container.
3. My first thought was to use `alignSelfCenter`, and that works. BUT at small screen widths, that style also centers the content vertically 😡 and adds a lot of unintended margin to the top and bottom
4. So I went with the auto margin route, which makes the form horizontally centered AND doesn't mess up any vertical margins

I have fiddled with it again today to remove the margin auto and get the `alignSelfCenter` working.

I narrowed the problem you saw with iPhone SE down to the `props.welcomeText`. It seems the difference in text was causing the width of the container to change size... so I have restricted that text to 300px wide always.

Updated!

shawnborton replied:
Could you (or a reviewer) share a quick video of the sign in page going from small to wide? Just want to check and see how the responsive styles are working, thanks!

tgolen replied:
Here you go:

https://user-images.githubusercontent.com/1228807/205146328-bdea7a15-d4f3-49bb-acb1-c06b65f14e92.mp4

shawnborton replied:
Cool, looks good to me, thanks for making the video!

tgolen committed:
Merge branch 'main' into tgolen-fix-signin-keyboard-avoiding

tgolen committed:
Merge branch 'main' into tgolen-fix-signin-keyboard-avoiding

tgolen commented:
Updated to resolve conflicts.

jasperhuangg reviewed (approved):
Code looks great and tests well on all platforms! Unable to replicate the broken small-screen android case, but I imagine that's an edge case so I'm approving. Let's create a follow-up issue for that?

tgolen commented:
Yep, since it's been twice approved, I'll go ahead and self-merge this so we can finally get it done.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3689736386) to production by @chiragsalian in version: 1.2.38-6 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

chiragsalian commented:
Just FYI that this PR caused https://github.com/Expensify/App/issues/13488.

eVoloshchak commented:
Leaving a note that this PR caused https://github.com/Expensify/App/issues/13920

eVoloshchak commented:
@rushatgabhane, sure!
On [this](https://github.com/Expensify/App/blob/bbba5e5bfd867f60ccc9c9f72f4f3eef12fc2f4f/src/pages/signin/SignInPageLayout/SignInPageContent.js#L41) line the sign-in page is wrapped with `TouchableWithoutFeedback`. It seems like there is an issue with RNPickerSelect when you enclose in with touchable on Android. It seems like it was enclosed with touchable to fix the issue specific to iOS 
```
 * On iOS sometimes the keyboard doesn't dismiss itself
 * when you tap outside of the input that pulled it up.
 * Therefore this component is necessary to ensure users
 * can still dismiss the keyboard on iOS.
```

So it was resolved by using platform-specific files and using touchable only on iOS
